calico/felix/fetcd.py

```diff
@@ -181,28 +182,44 @@ def watch_etcd(self):
                                                 wait=True,
                                                 waitIndex=next_etcd_index,
                                                 recursive=True,
-                                                timeout=0)
+                                                timeout=Timeout(connect=10,
+                                                                read=90))
                     _log.debug("etcd response: %r", response)
                 except ReadTimeoutError:
-                    _log.warning("Read from etcd timed out, retrying.")
-                    # TODO: We are timing out 60 seconds after the initial
-                    # read, then again 60 seconds after that, for no reason I
-                    # can really see. However, a timeout of 0 is probably wrong
-                    # if that does not reestablish connections
-                    # periodically. Needs a bit more thought.
+                    # This is expected when we're doing a poll and nothing
+                    # happened.
+                    _log.debug("Read from etcd timed out, retrying.")
                     continue
-                except EtcdException:
-                    _log.exception("Failed to read from etcd. wait_index=%s",
-                                   next_etcd_index)
-                    raise
-                # Defensive: ensure that, no matter what, we always
-                # monotonically increase the index we poll on.
-                if response.modifiedIndex < next_etcd_index:
-                    _log.error("Modified index on response < value we polled "
-                               "on: %s < %s", response.modifiedIndex,
-                               next_etcd_index)
-                next_etcd_index = max(response.modifiedIndex,
-                                      next_etcd_index + 1)
+                except EtcdException as e:
+                    # Sadly, python-etcd doesn't have a clean exception
+                    # hierarchy; look at the message.  We only log the stack
+                    # trace for errors we're not expecting to avoid copious
+                    # log spam.
+                    if e.message == "No more machines in the cluster":
```

Comment:
Is it a good idea to test that the message is exactly this? Might be better to have something like `if 'no more machines' in e.message.lower()`.